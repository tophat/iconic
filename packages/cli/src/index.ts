import * as path from 'path'

import yargs from 'yargs'
import * as fs from 'fs-extra'
import convert from '@react-vector-graphics/core'

function run(src: string, dest: string): void {
    const files = fs.readdirSync(src)
    const filesContents = files.map((file): [string, string] => [
        file,
        fs.readFileSync(path.join(src, file), { encoding: 'utf-8' }),
    ])
    const results = filesContents.map(([file, contents]): object =>
        convert(file, contents),
    )
    for (const result of results) {
        for (const [file, contents] of Object.entries(result)) {
            const filePath = path.join(dest, file)
            fs.outputFileSync(filePath, contents)
        }
    }
}

const argv = yargs
    .usage('Usage: $0 -s [src] -d [dest]')
    .option('src', {
        alias: 's',
        describe: 'source folder',
        type: 'string',
    })
    .option('dest', {
        alias: 'd',
        describe: 'destination folder',
        type: 'string',
    })
    .demandOption(['s', 'd']).argv

const src = argv.s as string
const dest = argv.d as string
run(src, dest)
