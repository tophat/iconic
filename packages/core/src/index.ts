export { getLogger } from './logging'
export { run as default } from './main'
