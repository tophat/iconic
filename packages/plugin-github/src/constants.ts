const ns = (s: TemplateStringsArray): string => `github/${s}`

export const OPTIONS = {
    API: ns`api`,
    BASE: ns`base`,
    COMMIT_CREATE_PATTERN: ns`commitCreate`,
    COMMIT_DELETE_PATTERN: ns`commitDelete`,
    COMMIT_UPDATE_PATTERN: ns`commitUpdate`,
    FILE_EXT: ns`fileExt`,
    FOLDER_PATH: ns`folderPath`,
    GLOB_PATTERN: ns`globPattern`,
    HEAD: ns`head`,
    NAME_SCHEME: ns`nameScheme`,
    OUTPUT_PATH: ns`outputPath`,
    OWNER: ns`owner`,
    REPO: ns`repo`,
}

export const STATE = {
    COMPONENT_FILES: ns`componentFiles`,
    COMPONENT_NAME: 'componentName', // @svgr/plugin-jsx
    COMPONENT_NAME_OLD: ns`componentNameOld`,
    DIFF_TYPE: ns`diffType`,
    FILE_PATH: 'filePath', // @svgr/plugin-jsx
}

export const STATUSES = {
    ADDED: 'added',
    MODIFIED: 'modified',
    REMOVED: 'removed',
    RENAMED: 'renamed',
}

export const COMMIT_MESSAGE_PLACEHOLDER = '${message}'

export const COMMIT_MESSAGE_PATTERNS = {
    CREATE: `feat: ${COMMIT_MESSAGE_PLACEHOLDER}`,
    DELETE: `refactor: ${COMMIT_MESSAGE_PLACEHOLDER}\nBREAKING CHANGE: ${COMMIT_MESSAGE_PLACEHOLDER}`,
    UPDATE: `fix: ${COMMIT_MESSAGE_PLACEHOLDER}`,
}
