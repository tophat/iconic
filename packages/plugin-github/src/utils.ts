import { Logger } from '@react-vector-graphics/types'

export const fromBase64 = (data: string): string =>
    Buffer.from(data, 'base64').toString()

export const toBase64 = (data: string): string =>
    Buffer.from(data).toString('base64')

export const eagerPromises = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    promises: Promise<any>[],
    logger: Logger = console,
): // eslint-disable-next-line @typescript-eslint/no-explicit-any
Promise<any> => Promise.all(promises.map(p => p.catch(logger.error)))
