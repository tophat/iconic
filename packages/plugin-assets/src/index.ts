import * as path from 'path'

import * as fs from 'fs-extra'
import * as glob from 'glob'

import {
    NamingScheme,
    Plugin,
    PluginParams,
} from '@react-vector-graphics/types'

import { pathToName } from './utils'

const ns = (s: TemplateStringsArray): string => `assets/${s}`

export const OPTIONS = {
    FILE_EXT: ns`fileExt`,
    GLOB_PATTERN: ns`globPattern`,
    NAME_SCHEME: ns`nameScheme`,
    OUTPUT_PATH: ns`outputPath`,
}

const run: Plugin = async (code, config, state) => {
    if (code) {
        const outputPath = config.options[OPTIONS.OUTPUT_PATH]
        if (outputPath && state.componentName) {
            const pathToFile = path.join(outputPath, state.componentName)
            const fileExt = config.options[OPTIONS.FILE_EXT]
            const filePath = fileExt ? `${pathToFile}.${fileExt}` : pathToFile
            fs.outputFileSync(filePath, code)
        }
        return { code, state }
    } else {
        const globPattern: string = config.options[OPTIONS.GLOB_PATTERN]
        const nameScheme: NamingScheme = config.options[OPTIONS.NAME_SCHEME]
        return glob.sync(globPattern).map(
            (filePath: string): PluginParams => ({
                code: fs.readFileSync(filePath, { encoding: 'utf-8' }),
                state: Object.assign(
                    {
                        componentName: pathToName(filePath, nameScheme),
                        filePath,
                    },
                    state,
                ),
            }),
        )
    }
}

export default run
