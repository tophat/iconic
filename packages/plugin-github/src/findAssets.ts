import * as path from 'path'

import minimatch from 'minimatch'
import { Octokit } from '@octokit/rest'

import { NamingScheme, PluginParams, State } from '@react-vector-graphics/types'
import { pathToName } from '@react-vector-graphics/utils'

import { STATE } from './constants'
import { fromBase64, normaliseGlob } from './utils'

const findAssets = async ({
    folderPath = '',
    github: { api: githubApi, ...githubParams },
    ...params
}: {
    folderPath: string
    github: {
        api: Octokit
        base: string
        head: string
        owner: string
        repo: string
    }
    globPattern: string
    nameScheme: NamingScheme
    state: State
}): Promise<PluginParams[]> => {
    const compareCommitsResult = await githubApi.repos.compareCommits(
        githubParams,
    )
    const svgFiles = compareCommitsResult.data.files.filter(file => {
        const isInFolder = file.filename.startsWith(folderPath)
        if (!isInFolder) return false
        const relPath = path.relative(folderPath, file.filename)
        return minimatch(relPath, normaliseGlob(params.globPattern))
    })
    const pluginParams = svgFiles.map(
        async (file): Promise<PluginParams> => {
            const filePath = path.relative(folderPath, file.filename)
            const { data } = await githubApi.repos.getContents({
                ...githubParams,
                path: file.filename,
                ref: githubParams.head,
            })
            if (Array.isArray(data) || !data.content) {
                throw new Error(`Could not get contents for ${file.filename}`)
            }
            // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
            // @ts-ignore
            const { previous_filename: previousFilename } = file
            return {
                code: fromBase64(data.content.toString()),
                state: Object.assign(
                    {
                        [STATE.COMPONENT_NAME]: pathToName(
                            filePath,
                            params.nameScheme,
                        ),
                        [STATE.COMPONENT_NAME_OLD]:
                            previousFilename &&
                            pathToName(
                                path.relative(folderPath, previousFilename),
                                params.nameScheme,
                            ),
                        [STATE.DIFF_TYPE]: file.status,
                        [STATE.FILE_PATH]: filePath,
                    },
                    params.state,
                ),
            }
        },
    )
    return Promise.all(pluginParams)
}

export default findAssets
