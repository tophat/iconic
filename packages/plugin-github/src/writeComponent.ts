import * as path from 'path'
import { promises } from 'dns'

import { Octokit } from '@octokit/rest'

import { Logger } from '@react-vector-graphics/types'

import {
    COMMIT_MESSAGE_PATTERNS,
    COMMIT_MESSAGE_PLACEHOLDER,
    OPTIONS,
    STATE,
    STATUSES,
} from './constants'
import { eagerPromises, replaceAll, toBase64 } from './utils'

const removeIconFiles = async (
    githubApi: Octokit,
    githubParams: { head: string; owner: string; repo: string },
    componentName: string,
    componentPath: string,
    commitMessagePattern: string = COMMIT_MESSAGE_PATTERNS.DELETE,
): Promise<void> => {
    const deleteMessage = `remove ${componentName}`
    const message = replaceAll(
        commitMessagePattern,
        COMMIT_MESSAGE_PLACEHOLDER,
        deleteMessage,
    )
    const { data: results } = await githubApi.repos.getContents({
        ...githubParams,
        path: componentPath,
        ref: githubParams.head,
    })
    await eagerPromises(
        (Array.isArray(results) ? results : [results]).map(({ path, sha }) =>
            githubApi.repos.deleteFile({
                ...githubParams,
                message,
                path,
                sha,
            }),
        ),
    )
}

const addOrModifyIconFile = async (
    githubApi: Octokit,
    githubParams: { head: string; owner: string; repo: string },
    componentName: string,
    fileName: string,
    filePath: string,
    fileContents: string,
    commitMessagePatternCreate: string = COMMIT_MESSAGE_PATTERNS.CREATE,
    commitMessagePatternUpdate: string = COMMIT_MESSAGE_PATTERNS.UPDATE,
): Promise<void> => {
    let fileSha
    try {
        const { data } = await githubApi.repos.getContents({
            ...githubParams,
            path: filePath,
            ref: githubParams.head,
        })
        if (Array.isArray(data)) return
        fileSha = data.sha
    } catch (e) {
        // assume file does not exist and do nothing
    }
    const message = fileSha
        ? replaceAll(
              commitMessagePatternUpdate,
              COMMIT_MESSAGE_PLACEHOLDER,
              `modify ${componentName} ${fileName}`,
          )
        : replaceAll(
              commitMessagePatternCreate,
              COMMIT_MESSAGE_PLACEHOLDER,
              `add ${componentName} ${fileName}`,
          )
    await githubApi.repos.createOrUpdateFile({
        ...githubParams,
        branch: githubParams.head,
        content: toBase64(fileContents),
        message,
        path: filePath,
        sha: fileSha,
    })
}

const writeComponent = async ({
    github: { api: githubApi, ...githubParams },
    ...params
}: {
    assetFile: string
    code: string
    commitMessagePatterns?: {
        create?: string
        delete?: string
        update?: string
    }
    componentName?: string
    componentNameOld?: string
    componentFiles: { [fileName: string]: string }
    diffType: string
    fileExt?: string
    folderPath: string
    github: {
        api: Octokit
        base: string
        head: string
        owner: string
        repo: string
    }
    logger?: Logger
    outputPath?: string
}): Promise<void> => {
    if (!params.componentName) {
        return params.logger?.warn(
            `No '${STATE.COMPONENT_NAME}' provided for '${params.assetFile}'.`,
        )
    }
    if (!params.outputPath) {
        return params.logger?.warn(`No '${OPTIONS.OUTPUT_PATH}' provided.`)
    }
    if (!params.fileExt) {
        params.logger?.warn(`No '${OPTIONS.FILE_EXT}' provided.`)
    }
    // gather files
    const componentFiles = Object.entries(params.componentFiles)
    const singleFile = componentFiles.length === 0
    const pathToFolder = path.join(
        params.folderPath,
        params.outputPath,
        singleFile ? '' : params.componentName,
    )
    const componentFileName =
        (singleFile ? params.componentName : 'index') +
        (params.fileExt ? `.${params.fileExt}` : '')
    componentFiles.push([componentFileName, params.code])
    const componentFilePath = path.join(pathToFolder, componentFileName)
    // commit file changes
    const pendingPromises = []
    if (params.diffType === STATUSES.REMOVED) {
        pendingPromises.push(
            removeIconFiles(
                githubApi,
                githubParams,
                params.componentName,
                singleFile ? componentFilePath : pathToFolder,
                params.commitMessagePatterns?.delete,
            ),
        )
    } else {
        // added, modified or renamed
        for (const [fileName, fileContents] of componentFiles) {
            const filePath = path.join(pathToFolder, fileName)
            pendingPromises.push(
                addOrModifyIconFile(
                    githubApi,
                    githubParams,
                    params.componentName,
                    fileName,
                    filePath,
                    fileContents,
                    params.commitMessagePatterns?.create,
                    params.commitMessagePatterns?.update,
                ),
            )
        }
        if (params.diffType === STATUSES.RENAMED && params.componentNameOld) {
            const oldPathToFolder = path.join(
                params.folderPath,
                params.outputPath,
                singleFile ? '' : params.componentNameOld,
            )
            const oldPathToFile = path.join(
                oldPathToFolder,
                singleFile ? params.componentNameOld : 'index',
            )
            const oldComponentFilePath = params.fileExt
                ? `${oldPathToFile}.${params.fileExt}`
                : oldPathToFile
            pendingPromises.push(
                removeIconFiles(
                    githubApi,
                    githubParams,
                    params.componentNameOld,
                    singleFile ? oldComponentFilePath : oldPathToFolder,
                    params.commitMessagePatterns?.delete,
                ),
            )
        }
    }
    await eagerPromises(pendingPromises)
}

export default writeComponent
