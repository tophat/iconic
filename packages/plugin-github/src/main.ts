import { Plugin } from '@react-vector-graphics/types'

import { OPTIONS, STATE } from './constants'
import findAssets from './findAssets'
import writeComponent from './writeComponent'

export const run: Plugin = async (code, config, state, logger) => {
    const github = {
        api: config.options[OPTIONS.API],
        base: config.options[OPTIONS.BASE],
        head: config.options[OPTIONS.HEAD],
        owner: config.options[OPTIONS.OWNER],
        repo: config.options[OPTIONS.REPO],
    }
    const sharedArgs = {
        folderPath: config.options[OPTIONS.FOLDER_PATH],
        github,
    }
    if (code) {
        await writeComponent({
            ...sharedArgs,
            assetFile: state[STATE.FILE_PATH],
            code,
            commitMessagePatterns: {
                create: config.options[OPTIONS.COMMIT_CREATE_PATTERN],
                delete: config.options[OPTIONS.COMMIT_DELETE_PATTERN],
                update: config.options[OPTIONS.COMMIT_UPDATE_PATTERN],
            },
            componentFiles: state[STATE.COMPONENT_FILES] ?? {},
            componentName: state[STATE.COMPONENT_NAME],
            componentNameOld: state[STATE.COMPONENT_NAME_OLD],
            diffType: state[STATE.DIFF_TYPE],
            fileExt: config.options[OPTIONS.FILE_EXT],
            logger,
            outputPath: config.options[OPTIONS.OUTPUT_PATH],
        })
        return { code, state }
    } else {
        return findAssets({
            ...sharedArgs,
            globPattern: config.options[OPTIONS.GLOB_PATTERN],
            nameScheme: config.options[OPTIONS.NAME_SCHEME],
            state,
        })
    }
}
