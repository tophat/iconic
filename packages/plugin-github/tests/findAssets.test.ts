import { Octokit } from '@octokit/rest'

import { NamingScheme } from '@react-vector-graphics/types'

import findAssets from '../src/findAssets'
import { OPTIONS } from '../src'

import { mockGithubApi, mockOptions } from './mocks'

describe('findAssets', () => {
    it('returns all assets matching pattern', async () => {
        await expect(
            findAssets({
                folderPath: mockOptions[OPTIONS.FOLDER_PATH] as string,
                github: {
                    api: (mockGithubApi as unknown) as Octokit,
                    base: mockOptions[OPTIONS.BASE] as string,
                    head: mockOptions[OPTIONS.HEAD] as string,
                    owner: mockOptions[OPTIONS.OWNER] as string,
                    repo: mockOptions[OPTIONS.REPO] as string,
                },
                globPattern: mockOptions[OPTIONS.GLOB_PATTERN] as string,
                nameScheme: mockOptions[OPTIONS.NAME_SCHEME] as NamingScheme,
                state: {},
            }),
        ).resolves.toMatchSnapshot()
    })
})
