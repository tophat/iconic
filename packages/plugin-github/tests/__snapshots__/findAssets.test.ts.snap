// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findAssets returns all assets matching pattern 1`] = `
Array [
  Object {
    "code": "<svg>mock</svg>",
    "state": Object {
      "componentName": "mockIcon3",
      "filePath": "assets/mock-icon-3.svg",
      "github/componentNameOld": undefined,
      "github/diffType": "added",
    },
  },
  Object {
    "code": "<svg>mock</svg>",
    "state": Object {
      "componentName": "mockIcon2",
      "filePath": "assets/mock-icon-2.svg",
      "github/componentNameOld": undefined,
      "github/diffType": "modified",
    },
  },
  Object {
    "code": "<svg></svg>",
    "state": Object {
      "componentName": "mockIcon",
      "filePath": "assets/mock.icon.svg",
      "github/componentNameOld": undefined,
      "github/diffType": "removed",
    },
  },
  Object {
    "code": "<svg>mock</svg>",
    "state": Object {
      "componentName": "mockIcon1",
      "filePath": "assets/mock-icon-1.svg",
      "github/componentNameOld": "mockIcon",
      "github/diffType": "renamed",
    },
  },
]
`;
